### ISSUE TITLE
'already borrowed: BorrowMutError'

### ISSUE THREAD

**Issue by glademiller** [sycamore-rs/sycamore#495]
**Describe the bug**
I have a view that contains an if statement that is based on a signal depending on the signals value a different subview is created. That subview has an input that is bound to a  new signal created within that view. Also it creates an effect based on that new signal. When that effect is run it attempts to change the value of the outer signal which triggers the BorrowMutError. My more complex example also sometimes simply stops handling the effect after the first event but I have provided a minimal example that always appears to work as stated here.

**To Reproduce**
```
use sycamore::prelude::*;

#[component]
pub fn TestPage<G: Html>(cx: Scope<'_>) -> View<G> {
    let signal = create_signal(cx, "test".to_string());
    view! { cx,
        div {
            "TEST PAGE"
            (if *signal.get() == "test" {
                let signalt = create_signal(cx, "test2".to_string());
                create_effect(cx, || {
                    signal.set(signalt.get().to_string());
                });
                view!{cx,
                    input(bind:value=signalt)
                }
            } else {
                let signals = create_signal(cx, "test3".to_string());
                create_effect(cx, || {
                    signal.set(signals.get().to_string());
                });
                view!{cx,
                    input(bind:value=signals)

                }
            })
        }
    }
}
```

**Expected behavior**
It should allow the flow above or perhaps something could be added to detect this at runtime if this isn't something that should work.

**Screenshots**
![image](https://user-images.githubusercontent.com/2006148/194343879-0e22705b-62b1-46e8-ac9e-f7e8ec822bb3.png)
![image](https://user-images.githubusercontent.com/2006148/194344015-a2952d4e-be69-42b7-b259-9199438231fe.png)

**Environment**

- Sycamore: 0.8 and master
- Browser: brave(chromium)
- OS: Linux

**Additional context**

**Comment by birktj** [comment]
I think I have come across the same bug. Here is an even smaller example, pressing the button instantly causes a `BorrowMutError` panic.

```rust
use sycamore::prelude::*;

fn main() {
    std::panic::set_hook(Box::new(console_error_panic_hook::hook));

    sycamore::futures::spawn_local(async {
        sycamore::render(|cx| {
            let a = create_signal(cx, true);

            view! {cx,
                (if *a.get() {
                    let b = create_signal(cx, ());
                    view!{cx,
                        button(on:click = move |_| {
                            a.set(false);
                            b.set(());
                        }) { "Button" }
                    }
                } else { view!{cx,} })
            }
        });
    });
}
```

**Comment by lukechu10** [comment]
Was hoping that this would get fixed by #626 but sadly it still persists.

**Comment by lukechu10** [comment]
The reason for the error is because the inner signal is being accessed after it has been destroyed. This is because when setting the outer signal, the inner scope is rerun which destroys the inner signal.

The solution to this is to access the inner signal first, and then set the outer signal:
```rust
let inner = inner_signal.get();
outer_signal.set(f(inner));
```

This is still an error in the latest version (`0.9.0-beta.3`) but should be a bit clearer since we get a `signal was disposed` error pointing to the location where the signal was originally created.
